File: com.chartboost.helium/Android~/helium-android-bridge/src/main/java/com/chartboost/heliumsdk/unity/HeliumEventProcessor.java
package com.chartboost.heliumsdk.unity;

import android.util.Log;
import com.chartboost.heliumsdk.ad.HeliumAdError;
import org.json.JSONException;
import org.json.JSONObject;
import java.util.HashMap;

public class HeliumEventProcessor {
    private static final String TAG = "HeliumEventProcessor";
    private static final String EMPTY_STRING = "";

    @FunctionalInterface
    public interface HeliumEventConsumer<T, V, S> {
        void accept(T placementName, V errorCode, S errorDescription);
    }

    public static void serializeHeliumEvent(String placementName, HeliumAdError heliumAdError, HeliumEventConsumer<String, Integer, String> eventConsumer){
        if (placementName == null)
            placementName = EMPTY_STRING;

        int errorCode = -1;
        String errorDescription = EMPTY_STRING;

        if (heliumAdError != null) {
            int tempCode = heliumAdError.getCode();
            errorDescription = heliumAdError.getMessage();

            // we do this in order to bypass lint issues
            if (errorCode != tempCode)
                errorCode = tempCode;
        }

        eventConsumer.accept(placementName, errorCode, errorDescription);
    }

    @FunctionalInterface
    public interface HeliumBidEventConsumer<T, V, S, X>
    {
        void accept(T placementName, V auctionId, S partnerId, X price);
    }

    public static void serializeHeliumBidEvent(String placementName, HashMap<String, String> dataMap, HeliumBidEventConsumer<String, String, String, Double> eventConsumer)
    {
        try {
            if (placementName == null)
                placementName = EMPTY_STRING;

            String partnerId = dataMap.get("partner_id");
            partnerId = partnerId == null ? EMPTY_STRING : partnerId;
            String auctionId = dataMap.get("auction-id");
            auctionId = auctionId == null ? EMPTY_STRING : auctionId;
            String priceAsString = dataMap.get("price");
            priceAsString = priceAsString == null ? "0" : priceAsString;
            double price = Double.parseDouble(priceAsString);
            eventConsumer.accept(placementName, auctionId, partnerId, price);
        }
        catch (Exception e) {
            Log.d(TAG, "bidFetchingInformationError", e);
        }
    }

    @FunctionalInterface
    public interface HeliumRewardEventConsumer<T, V>
    {
        void accept(T placementName, V reward);
    }

    public static void serializeHeliumRewardEvent(String placementName, String reward, HeliumRewardEventConsumer<String, Integer> eventConsumer) {
        if (placementName == null)
            placementName = EMPTY_STRING;

        int rewardAsInt = 0;
        if (reward != null)
            rewardAsInt = Integer.parseInt(reward);

        eventConsumer.accept(placementName, rewardAsInt);
    }

    public static String serializePlacementILRDData(String placementName, JSONObject ilrdInfo) {
        JSONObject serializedString = new JSONObject();
        try {
            serializedString.put("placementName", placementName);
            serializedString.put("ilrd", ilrdInfo);
        } catch (JSONException e) {
            Log.d(TAG, "serializeError", e);
        }
        return serializedString.toString();
    }
}

File: com.chartboost.helium/Android~/helium-android-bridge/src/main/kotlin/com/chartboost/heliumsdk/unity/HeliumEventProcessor.kt
package com.chartboost.heliumsdk.unity

import android.util.Log
import com.chartboost.heliumsdk.ad.HeliumAdError
import org.json.JSONException
import org.json.JSONObject

object HeliumEventProcessor {
    private const val TAG = "HeliumEventProcessor"
    private const val EMPTY_STRING = ""

    @JvmStatic
    fun serializeHeliumEvent(
        placement: String?,
        heliumAdError: HeliumAdError?,
        eventConsumer: HeliumEventConsumer<String?, Int, String?>
    ) {
        var placementName: String? = EMPTY_STRING
        if (placement != null) placementName = placement
        var errorCode = -1
        var errorDescription = EMPTY_STRING
        if (heliumAdError != null) {
            val tempCode = heliumAdError.getCode()
            errorDescription = heliumAdError.getMessage()

            // we do this in order to bypass lint issues
            if (errorCode != tempCode) errorCode = tempCode
        }
        eventConsumer.accept(placementName, errorCode, errorDescription)
    }

    @JvmStatic
    fun serializeHeliumBidEvent(
        placement: String?,
        dataMap: HashMap<String, String>,
        eventConsumer: HeliumBidEventConsumer<String?, String?, String?, Double>
    ) {
        try {
            var placementName: String? = EMPTY_STRING
            if (placement != null) placementName = placement
            var partnerId = dataMap["partner_id"]
            partnerId = partnerId ?: EMPTY_STRING
            var auctionId = dataMap["auction-id"]
            auctionId = auctionId ?: EMPTY_STRING
            var priceAsString = dataMap["price"]
            priceAsString = priceAsString ?: "0"
            val price = priceAsString.toDouble()
            eventConsumer.accept(placementName, auctionId, partnerId, price)
        } catch (e: Exception) {
            Log.d(TAG, "bidFetchingInformationError", e)
        }
    }

    @JvmStatic
    fun serializeHeliumRewardEvent(
        placement: String?,
        reward: String?,
        eventConsumer: HeliumRewardEventConsumer<String?, Int>
    ) {
        var placementName: String? = EMPTY_STRING
        if (placement != null) placementName = placement
        try {
            var rewardAsInt = 0
            if (reward != null) rewardAsInt = reward.toInt()
            eventConsumer.accept(placementName, rewardAsInt)
        } catch (exception: Exception) {
            Log.e(TAG, "Failed to Parse Reward Information: Reward: $reward")
            eventConsumer.accept(placementName, 1)
        }
    }

    @JvmStatic
    fun serializePlacementILRDData(placementName: String?, ilrdInfo: JSONObject?): String {
        val serializedString = JSONObject()
        try {
            serializedString.put("placementName", placementName)
            serializedString.put("ilrd", ilrdInfo)
        } catch (e: JSONException) {
            Log.d(TAG, "serializeError", e)
        }
        return serializedString.toString()
    }

    fun interface HeliumEventConsumer<T, V, S> {
        fun accept(placementName: T, errorCode: V, errorDescription: S)
    }

    fun interface HeliumBidEventConsumer<T, V, S, X> {
        fun accept(placementName: T, auctionId: V, partnerId: S, price: X)
    }

    fun interface HeliumRewardEventConsumer<T, V> {
        fun accept(placementName: T, reward: V)
    }
}
